src/olmo_core/nn/rope.py

```diff
@@ -43,32 +43,191 @@ class RoPEType(StrEnum):
 @dataclass
 class RoPEScalingConfig(Config):
     """
-    Defines how to scale RoPE to longer sequence lengths.
+    Base class for RoPE scaling configs. Defines a strategy for scaling RoPE to longer sequences.
+    """
+
+    attention_rescale_factor: float = 1.0
+    """
+    Factor to rescale attention scores by when using scaled RoPE. Can be used to compensate for
+    the larger effective context.
+    """
+
+    def __post_init__(self):
+        if self.attention_rescale_factor < 1.0:
+            raise OLMoConfigurationError(
+                "attention_factor must be >= 1.0, but got {self.attention_factor}"
+            )
+
+    @abstractmethod
+    def compute_scaled_inv_freq(
+        self, theta: int, dim: int, device: torch.device
+    ) -> tuple["torch.Tensor", float]:
+        """Compute the scaled inverse frequencies for RoPE, and the attention rescaling factor."""
+        raise NotImplementedError
+
+    @staticmethod
+    def compute_inv_freqs(theta: int, dim: int, device: torch.device) -> "torch.Tensor":
+        inv_freq = 1.0 / (
+            theta ** (torch.arange(0, dim, 2, device=device, dtype=torch.float) / dim)
+        )
+        return inv_freq
+
+
+@dataclass
+class PIRopeScalingConfig(RoPEScalingConfig):
+    """
+    Position-Interpolation (PI) RoPE scaling from Chen et al. (https://arxiv.org/pdf/2306.15595)
+
+    Interpolate the rotary angles instead of extrapolating them when the context window at
+    inference time exceeds the window used during training. In practice, this amounts to linearly
+    *compressing* the original position indices by a constant factor ``factor``.
+    """
+
+    factor: float = 2.0
+    """Context expansion multiplier. If factor = 1, reduces to vanilla RoPE."""
+
+    def compute_scaled_inv_freq(
+        self, theta: int, dim: int, device: torch.device
+    ) -> tuple["torch.Tensor", float]:
+        inv_freq = self.compute_inv_freqs(theta, dim, device)
+
+        # Positional-interpolation scales the *positions* by 1/factor. This is
+        # equivalent to scaling the inverse frequencies by the same amount.
+        if self.factor != 1.0:
+            inv_freq = inv_freq / self.factor
+
+        return inv_freq, self.attention_rescale_factor
+
+
+@dataclass
+class PerFrequencyRoPEScalingConfig(RoPEScalingConfig):
```

Comment:
nit, but the name suggests to me that you get more finegrained control than high/low...